FuzzyART/FuzzyArt/FuzzyART/FuzzyART.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 09/08/14               \" DATE 
.Dt FuzzyART 1      \" Program name and manual section number 
.Os Darwin
.Sh NAME                 \" Section Header - required - don't modify 
.Nm FuzzyART,
.\" The following lines are read in generating the apropos(man -k) database. Use only key
.\" words here as the database is built based on the words here and in the .ND line. 
.Nm Other_name_for_same_program(),
.Nm Yet another name for the same program.
.\" Use .Nm macro to designate other names for the documented program.
.Nd This line parsed for whatis database.
.Sh SYNOPSIS             \" Section Header - required - don't modify
.Nm
.Op Fl abcd              \" [-abcd]
.Op Fl a Ar path         \" [-a path] 
.Op Ar file              \" [file]
.Op Ar                   \" [file ...]
.Ar arg0                 \" Underlined argument - use .Ar anywhere to underline
arg2 ...                 \" Arguments
.Sh DESCRIPTION          \" Section Header - required - don't modify
Use the .Nm macro to refer to your program throughout the man page like such:
.Nm
Underlining is accomplished with the .Ar macro like this:
.Ar underlined text .
.Pp                      \" Inserts a space
A list of items with descriptions:
.Bl -tag -width -indent  \" Begins a tagged list 
.It item a               \" Each item preceded by .It macro
Description of item a
.It item b
Description of item b
.El                      \" Ends the list
.Pp
A list of flags and their descriptions:
.Bl -tag -width -indent  \" Differs from above in tag removed 
.It Fl a                 \"-a flag as a list item
Description of -a flag
.It Fl b
Description of -b flag
.El                      \" Ends the list
.Pp
.\" .Sh ENVIRONMENT      \" May not be needed
.\" .Bl -tag -width "ENV_VAR_1" -indent \" ENV_VAR_1 is width of the string ENV_VAR_1
.\" .It Ev ENV_VAR_1
.\" Description of ENV_VAR_1
.\" .It Ev ENV_VAR_2
.\" Description of ENV_VAR_2
.\" .El                      
.Sh FILES                \" File used or created by the topic of the man page
.Bl -tag -width "/Users/joeuser/Library/really_long_file_name" -compact
.It Pa /usr/share/file_name
FILE_1 description
.It Pa /Users/joeuser/Library/really_long_file_name
FILE_2 description
.El                      \" Ends the list
.\" .Sh DIAGNOSTICS       \" May not be needed
.\" .Bl -diag
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .El
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr a 1 , 
.Xr b 1 ,
.Xr c 1 ,
.Xr a 2 ,
.Xr b 2 ,
.Xr a 3 ,
.Xr b 3 
.\" .Sh BUGS              \" Document known, unremedied bugs 
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner

GRT-FuzzyART/GRT-FuzzyART/GRT-FuzzyART/FuzzyART.cpp
//
//  FuzztART.cpp
//  GRT-FuzzyART
//
//  Created by Chikashi Miyama on 11/08/14.
//  Copyright (c) 2014 ICST. All rights reserved.
//

#include "FuzzyART.h"

using namespace GRT;

RegisterClustererModule<FuzzyART> FuzzyART::registerModule("FuzzyART");

FuzzyART::FuzzyART(const UINT choice, double learnRate, double vigilance){
    
    this->choice = choice;
    this->learnRate = learnRate;
    this->vigilance = vigilance;
    
}

FuzzyART::~FuzzyART(){
    
}

bool FuzzyART::reset(){
    residual = 0;
    dimensions = 3;
    inputCount = 0;
    
    inputVector.clear();
    inputVector.resize(dimensions*2); // because of complemented code
    
    recentChoice = -1;
    acVector.clear();
    acVector.push_back(ArtCategory(dimensions));
    
    return true;
}

bool FuzzyART::train_(MatrixDouble &data){
    
    UINT numRows = data.getNumRows();
    labels.clear();
    
    for(int i = numRows; i < numRows; i++){
        std::vector<double> inv = data.getRowVector(i);
        setInput(inv); // copy input vector to the buffer, normalize, and prepare the complement code
        fillCategoryChoice(); // check the distance between input and each cateory
        makeChoice(vigilance); // make final prediction, update recent Choice, if the input is far from all existing AC make new one, depending on vigilance
    
        labels.push_back(recentChoice);
    }
    
    return true; // return result
}



// process input vector
void FuzzyART::setInput(std::vector<double> inv){ // input is without complemented code
    
    dimensions = static_cast<int>(inv.size()); // without complemented code
    
    // 1. resize the dimension of all ACs if necessary
    std::vector<ArtCategory>::iterator ait = acVector.begin();
    while(ait != acVector.end()){
        (*ait).resize(dimensions);
        ait++;
    }
    
    // 2. normalize inv and copy it to inputVector with complement code
    inputVector.clear();
    double max;
    getMax(inv, max);
    std::vector<double>::iterator it = inv.begin();
    while (it != inv.end()) {
        double normalized = (*it)/max;
        inputVector.push_back(normalized);
        inputVector.push_back(1.0-normalized);
        it++;
    }
}

// calculate the distances between each category and input vector and store the result in choices vector
void FuzzyART::fillCategoryChoice(){
    
    choices.clear();
    choices.resize(acVector.size()); // number of choices is depending on the number of ART Categories
    
    // check against all existing categories, and 1 empty one
    for(int i = 0; i < choices.size(); i++){
        choices[i] = acVector[i].choose(inputVector, choice);
    }
}

// search vector, return max index and put max value in the specified buffer
int FuzzyART::getMax(std::vector<double> vec, double &maxValue){
    int maxIndex = -1;
    double max = 0;
    for (int i = 0; i < vec.size(); i++){
        if (vec[i] > max){
            max = vec[i];
            maxIndex = i;
        }
    }
    maxValue = max;
    return maxIndex;
}

int FuzzyART::makeChoice(double vig){
    int maxIndex = -1;
    bool chosen = false; // check mVigilance stuff here...
    while (!chosen){
        
        // 1. find largest match value in the vector of choices
        double max = 0;
        maxIndex = getMax(choices, max);
        
        // 2. if something is found
        if (maxIndex != -1){
            
            // 2-1 get a pointer to AC instance with the biggest match value (closest category)
            ArtCategory* ac = &acVector[maxIndex];
            
            // 2-2 if above vigilence (i.e. the closest one is not close enough)  or there is only one category, then learn
            
            if (ac->evalVigilance(inputVector, vig) || acVector.size() == 1){     // learn!
                residual = ac->learn(inputVector, learnRate); //learn
                while (maxIndex >= acVector.size()-1){  // committed the previous uncommitted category, so add a new blank one.
                    acVector.push_back(ArtCategory(dimensions));
                }
                chosen = true;
                recentChoice = maxIndex;
            }else{  // failed the mVigilance test.
                choices[maxIndex] = -1; // reset, try again
                maxIndex = -1;
            }
        } else {
            chosen = true;
        }
    } // otherwise look again.
    
    if (maxIndex > -1){
        inputCount++;
    }
    
    return maxIndex;
}



void FuzzyART::dump(){
    std::cout <<  "learnRate:" << learnRate << std::endl;
    std::cout <<  "vigilance:" << vigilance << std::endl;
    std::cout <<  "choice:" << choice << std::endl;
    std::cout <<  "input count:" << inputCount << std::endl;  // how many inputs we have seen
    
    std::vector<ArtCategory>::iterator it = acVector.begin();
    int index = 0;
    while(it != acVector.end()){
        std::cout <<  "======= category:" << index <<  "========" << std::endl;
        (*it).dump();
        it++;
        index++;
    }
}
//
//UINT FuzzyART::increaseVigilance(){
//    if (recentChoice > -1){
//        ArtCategory* ac = &acVector[recentChoice];
//        double higher_vig = ac->getVigilance(inputVector) + 0.01f;    // increase by a little bit.
//        recentChoice = makeChoice(higher_vig);
//    }
//    return recentChoice;
//}
